---
title: "Fig S3 - Human cohort 16S results"
author: "Jeff Letourneau"
date: '2022-04-27'
output: html_document
---

# Load libraries
```{r}
library(tidyverse)
library(phyloseq)
library(patchwork)
library(ggtext)
library(RColorBrewer)
library(vegan)
library(ALDEx2)

theme_set(theme_bw() +
            theme(axis.text = element_text(color = "black"),
                  axis.ticks = element_line(color = "black"),
                  plot.title = element_text(hjust = 0.5)))

```

# Load data
```{r}
set.seed(123)

# Human cohort 16S
ps <- readRDS("data/16S human cohort/phyloseq.rds")

# Adjust time labels
ps@sam_data <- ps@sam_data %>%
  as.matrix() %>%
  as.data.frame() %>%
  mutate(time = case_when(day == "T1" ~ -6,
                          day == "F1" ~ -3,
                          day == "T2" ~ 1,
                          day == "F2" ~ 4,
                          day == "T3" ~ 8,
                          day == "F3" ~ 11)) %>%
  sample_data()
ps@sam_data$time <- factor(ps@sam_data$time)
  
# Adjust taxa naming convention
## For ALDEx2, only examining taxa found in most participants
colnames(ps@otu_table) <- gsub("seq", "ASV", colnames(ps@otu_table))
rownames(ps@tax_table) <- colnames(ps@otu_table)

ps_subset <- ps %>%
  prune_samples(sample_sums(.) >= 5000, .) %>%
  filter_taxa(function(x) sum(x > 3) > 0.50*length(x), TRUE)

# Metabolic potential data
mp <- read.csv("data/HPAEC/human_cohort_HPAEC_all.csv") %>%
  filter(cond == "pump" & DPn >= 3) %>%
  mutate(group = ifelse(arm == "blue", "Prebiotic", "Placebo")) %>%
  group_by(participant, group, time) %>%
  summarize(total = sum(conc))

mp_F1 <- mp %>%
  filter(time == "F1")

mp_avg_change <- mp %>%
  filter(group == "Prebiotic") %>%
  spread(key=time, value=total) %>%
  mutate(change = mean(c(T2, F2), na.rm=T) / mean(c(T1, F1), na.rm=T))

```

# Calculate statistics
```{r}
# PERMANOVA by time
bray <- as.data.frame(as.matrix(distance(ps, method = "bray")))
adonis2(bray ~ group * time, data = as.data.frame(as.matrix(ps@sam_data)), strata = ps@sam_data$participant)
#             Df SumOfSqs      R2      F Pr(>F)  
# group        1    1.239 0.01769 4.1059  0.071 .
# time         5    0.976 0.01392 0.6465  0.053 .
# group:time   5    0.845 0.01205 0.5596  0.359  

# ALDEx2 GLM
samdf_subset <- as.data.frame(as.matrix(ps_subset@sam_data))
samdf_subset$time <- factor(samdf_subset$time, levels = c("-6", "-3", "1", "4", "8", "11"))

otu_flip <- as.data.frame(t(ps_subset@otu_table))
mm <- model.matrix(~ group*time, samdf_subset)

clr <- aldex.clr(otu_flip, mm, mc.samples=128, denom="all")
glm.test <- aldex.glm(clr)

hits_df <- glm.test %>%
  filter(`model.groupPrebiotic:time1 Pr(>|t|).BH` < 0.05 |
           `model.groupPrebiotic:time4 Pr(>|t|).BH` < 0.05 |
           `model.groupPrebiotic:time8 Pr(>|t|).BH` < 0.05 |
           `model.groupPrebiotic:time11 Pr(>|t|).BH` < 0.05) # no hits

# ALDEx2 correlation
## Does baseline taxonomy explain differences in baseline MP (all 40)?
ps_baseline <- ps %>%
  subset_samples(day == "F1") %>%
  prune_samples(sample_sums(.) >= 5000, .) %>%
  filter_taxa(function(x) sum(x > 3) > 0.50*length(x), TRUE)
otu_flip_mp <- as.data.frame(t(ps_baseline@otu_table))

# Match samples
colnames(otu_flip_mp) <- substr(colnames(otu_flip_mp), 1, 4)
otu_flip_mp <- otu_flip_mp[, colnames(otu_flip_mp) %in% mp_F1$participant]
mp_F1 <- mp_F1[mp_F1$participant %in% colnames(otu_flip_mp),]
otu_flip_mp <- otu_flip_mp[, mp_F1$participant]

# Run test
clr_mp <- aldex.clr(otu_flip_mp, mp_F1$total)
corr_test_mp <- aldex.corr(clr_mp, mp_F1$total)

## Does baseline (F1) taxonomy explain differences in MP change (prebiotic group only)?
ps_prebiotic <- ps %>%
  subset_samples(group == "Prebiotic" & day == "F1") %>%
  prune_samples(sample_sums(.) >= 5000, .) %>%
  filter_taxa(function(x) sum(x > 3) > 0.50*length(x), TRUE)

otu_flip_pre <- as.data.frame(t(ps_prebiotic@otu_table))

# Match samples
colnames(otu_flip_pre) <- substr(colnames(otu_flip_pre), 1, 4)
otu_flip_pre <- otu_flip_pre[, colnames(otu_flip_pre) %in% mp_avg_change$participant]
mp_avg_change <- mp_avg_change[mp_avg_change$participant %in% colnames(otu_flip_pre),]
otu_flip_pre <- otu_flip_pre[, mp_avg_change$participant]

# Run test
clr_pre <- aldex.clr(otu_flip_pre, mp_avg_change$change)
corr_test_pre <- aldex.corr(clr_pre, mp_avg_change$change)


## Does final (F2) taxonomy explain differences in MP change (prebiotic group only)?
ps_prebiotic2 <- ps %>%
  subset_samples(group == "Prebiotic" & day == "F2") %>%
  prune_samples(sample_sums(.) >= 5000, .) %>%
  filter_taxa(function(x) sum(x > 3) > 0.50*length(x), TRUE)

otu_flip_pre2 <- as.data.frame(t(ps_prebiotic2@otu_table))

# Match samples
mp_avg_change <- mp %>%
  filter(group == "Prebiotic") %>%
  spread(key=time, value=total) %>%
  mutate(change = mean(c(T2, F2), na.rm=T) / mean(c(T1, F1), na.rm=T))

colnames(otu_flip_pre2) <- substr(colnames(otu_flip_pre2), 1, 4)
otu_flip_pre2 <- otu_flip_pre2[, colnames(otu_flip_pre2) %in% mp_avg_change$participant]
mp_avg_change <- mp_avg_change[mp_avg_change$participant %in% colnames(otu_flip_pre2),]
otu_flip_pre2 <- otu_flip_pre2[, mp_avg_change$participant]

# Run test
clr_pre2 <- aldex.clr(otu_flip_pre2, mp_avg_change$change)
corr_test_pre2 <- aldex.corr(clr_pre2, mp_avg_change$change)




```

# Generate plots
```{r}
# Ordination plot by time point w/ PERMANOVA (strata = participant)
ord <- ordinate(ps, method="NMDS", distance="bray")
ord_plot <- plot_ordination(ps, ord, color="time", title="Bray-Curtis NMDS") +
  facet_wrap(~group) + scale_color_brewer(palette = "Paired") +
  stat_ellipse(type = "t") +
  labs(color = "Time (day)")

# ALDEx2 volcano plot
volcano_1 <- ggplot(glm.test, aes(x = `model.groupPrebiotic:time1 Estimate`,
                                y = -log10(`model.groupPrebiotic:time1 Pr(>|t|)`))) +
  geom_point() + 
  labs(x = "Effect size estimate", y = "-log<sub>10</sub>(*p*-value)") +
  theme(axis.title.y = element_markdown()) +
  ggtitle("ALDEx2 GLM\nPrebiotic group : day 1")

volcano_4 <- ggplot(glm.test, aes(x = `model.groupPrebiotic:time4 Estimate`,
                                y = -log10(`model.groupPrebiotic:time4 Pr(>|t|)`))) +
  geom_point() + 
  labs(x = "Effect size estimate", y = "-log<sub>10</sub>(*p*-value)") +
  theme(axis.title.y = element_markdown()) +
  ggtitle("ALDEx2 GLM\nPrebiotic group : day 4")

# Does baseline (F1) taxonomy correlate with baseline MP?
volcano_mp_baseline <- ggplot(corr_test_mp, aes(x=spearman.erho, y = -log10(spearman.eBH))) +
  geom_point() +
  labs(x = "Spearman's rho", y = "-log<sub>10</sub>(BH-corrected *p*-value)") +
  theme(axis.title.y = element_markdown()) +
  ggtitle("ALDEx2 correlation\nday -3 taxa vs. MP")


# Does taxonomy explain differences in MP change?
volcano_mp_pre <- ggplot(corr_test_pre, aes(x=spearman.erho, y = -log10(spearman.eBH))) +
  geom_point() +
  labs(x = "Spearman's rho", y = "-log<sub>10</sub>(BH-corrected *p*-value)") +
  theme(axis.title.y = element_markdown()) +
  ggtitle("ALDEx2 correlation\nday -3 taxa vs. ΔMP")

volcano_mp_pre2 <- ggplot(corr_test_pre2, aes(x=spearman.erho, y = -log10(spearman.eBH))) +
  geom_point() +
  labs(x = "Spearman's rho", y = "-log<sub>10</sub>(BH-corrected *p*-value)") +
  theme(axis.title.y = element_markdown()) +
  ggtitle("ALDEx2 correlation\nday 4 taxa vs. ΔMP")


```

# Random forest model
```{r}
library(randomForest)
library(caTools)
library(caret)
library(svMisc)
library(ROCR)
library(e1071)
library(pROC)
library(MLmetrics)
library(MLeval)

set.seed(12345) # Random seed, keep this to obtain the same figure as in the paper

# Calculate CLR change
## Melt phyloseq object
ps.clr <- prune_samples(sample_sums(ps) >= 5000, ps) %>%
  filter_taxa(function(x) sum(x > 3) > 0.50*length(x), TRUE) %>%
  microbiome::transform(., 'clr')
melty.clr <- psmelt(ps.clr)
length(unique(melty.clr$OTU)) # down to 89 taxa

# Calculate average change from week 1 to week 2
clr.change <- melty.clr[melty.clr$day %in% c("T1", "F1", "T2", "F2"), c(5,1,7,6,3)] %>%
  spread(key=day, value=Abundance)
clr.change$change <- NA
for(r in 1:nrow(clr.change)) {
  clr.change$change[r] <- mean(c(clr.change$T2[r], clr.change$F2[r]), na.rm=T) - mean(c(clr.change$T1[r], clr.change$F1[r]), na.rm=T)
}

clr.change <- clr.change %>%
  group_by(participant, group, OTU) %>%
  dplyr::summarize(mean=mean(change)) %>%
  spread(key=OTU, value=mean) %>%
  as.data.frame() %>%
  ungroup()
rownames(clr.change) <- clr.change$participant



# Properly format for random forest
clr <- clr.change[,c(3:91, 2)]


# Create empty objects to be loaded with output statistics
importanceDf <- data.frame()
accuracy <- c()
kappa <- c()
pvals <- c()
auc <- c()
roc.data <- data.frame()

# Iterate through the random forest model 100 times
for (i in 1:100){
  # create training/testing split
  data <- clr
  rows <- sample(nrow(clr)) # randomly shuffle rows of the dataset to create additional variation between iterations
  data <- data[rows,]
  
  # Train the model
  fitControl <- trainControl(method = "LOOCV",
                             summaryFunction = prSummary,
                             classProbs = T,
                             savePredictions = TRUE) 
  rf <- train(group ~ ., data = data,
              method = "rf",
              trControl = fitControl,
              tuneGrid = expand.grid(.mtry=sqrt(ncol(data)-1)), # default val of sqrt(# features)
              verbose = FALSE)
  
  # Store the scaled importance values
  importances <- varImp(rf)$importance %>% as.matrix %>% t()
  
  # Summarize confusion matrix
  confusion <- confusionMatrix(data = rf$pred$pred, reference = rf$pred$obs,
                               mode = "prec_recall", positive = "Prebiotic")
  
  # Compile resulting metrics
  ## Could pull other values (Mcnemar's Test P-Value, Precision, Recall, F1, etc.) in similar manner
  importanceDf <- rbind(importanceDf, importances)
  accuracy <- append(accuracy, confusion$overall["Accuracy"])
  kappa <- append(kappa, confusion$overall["Kappa"])
  pvals <- append(pvals, confusion$overall["AccuracyPValue"])
  
  #roc <- append(roc, auc(rf$pred$obs[selectedIndices], rf$pred$Placebo[selectedIndices]))
  
  # Compile data for plotting ROC curve
  plots <- evalm(rf)
  roc.data.tmp <- data.frame(SENS=plots$roc$data$SENS, FPR=plots$roc$data$FPR,
                             point=1:42, iteration=paste0("iteration", i))
  roc.data <- rbind(roc.data, roc.data.tmp)
  auc <- append(auc, plots$stdres$`Group 1`[13,1])

  # Print out how far through the iterations we are
  print(paste0(i, "/100 iterations complete"))
}

# Generate ROC curve
roc.avg <- roc.data[,1:3] %>%
  gather(key=variable, value=val, 1:2) %>%
  group_by(variable, point) %>%
  summarize(mean=mean(val)) %>%
  spread(key=variable, value=mean)

roc.plot <- ggplot(roc.data, aes(x=FPR, y=SENS, group=iteration)) +
  geom_line(color="red", alpha=0.2) +
  geom_line(data=roc.avg, aes(x=FPR, y=SENS), inherit.aes = F, color="black", size=1) +
  labs(x="False Positive Rate", y="True Positive Rate") +
  geom_abline(intercept = 0, slope = 1, color="darkgray", linetype = "dashed") +
  annotate(geom="text", x=0.7, y=0.25, label=paste0("Mean AUC = ", round(mean(auc), 3), "\n",
                                                     "Mean Accuracy = ", round(mean(accuracy), 3), "\n",
                                                     "Mean Kappa = ", round(mean(kappa),3), "\n",
                                                     "Mean p(Acc < NIR) = ", round(mean(pvals), 3)),
           size=2.8)


# calculate average importance across runs and then sort the data
avgImportances <- importanceDf %>% 
  colMeans() %>% 
  sort(decreasing = TRUE) %>%
  .[1:10] # select top 10 taxa

# link seq numbers to their family taxa name
taxa <- tax_table(ps.clr) %>% as.data.frame()
taxa <- taxa[names(avgImportances),]
taxa$Species[is.na(taxa$Species)] <- "sp."
taxa$binom <- paste0(taxa$Genus, " ", taxa$Species, " (", rownames(taxa), ")")

top10 <- data.frame(tax=names(avgImportances), importance=avgImportances, binom = taxa$binom)
top10$binom <- factor(top10$binom, levels=rev(top10$binom))

# Importance plot
importance.plot <- ggplot(top10, aes(x=importance, y=binom)) +
  geom_point() +
  theme_bw() +
  labs(x="Importance", y=NULL) +
  theme(axis.text.y=element_text(size=6))

# CLR plot
clr.select <- clr[,colnames(clr) %in% c("group", rownames(top10))]
clr.select <- cbind(data.frame(participant=rownames(clr.select)), clr.select[,c(11,1:10)])
clr.select <- clr.select %>% 
  gather(key=tax, value=clr, 3:12)
for(r in 1:nrow(clr.select)) {
  clr.select$binom[r] <- as.character(top10[top10$tax == clr.select$tax[r], "binom"])
}
clr.select$binom <- factor(clr.select$binom, levels=rev(top10$binom))

clr.summary <- clr.select %>%
  group_by(group, binom) %>%
  summarize(mean_clr=mean(clr), se=sd(clr)/sqrt(length(clr)))

clr.plot <- ggplot(clr.select, aes(x=clr, y=binom, group=group, color=group)) +
  geom_point(position=position_jitter(height=0.1), alpha=0.5, size=0.5) +
  geom_errorbar(data=clr.summary, aes(y=binom, xmin=mean_clr, xmax=mean_clr, group=group, color=group),
                size=0.75, width=0.6, inherit.aes = F) +
    geom_errorbar(data=clr.summary, aes(y=binom, xmin=mean_clr-se, xmax=mean_clr+se, group=group, color=group),
                size=0.75, width=0.3, inherit.aes = F) +
  theme_bw() +
  labs(x="ΔCLR(Relative Abundance)", y=NULL, color="Group") +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())

join.plot <- importance.plot + clr.plot + plot_layout(widths = c(1, 2))

```

# Random forest model again...
```{r}
# This time, subset on inulin degrading taxa

# Export taxa table and manually match with exported VMH inulin degraders
# write.csv(as.data.frame(tax_table(ps.clr)), "data/16S human cohort/taxa_filtered.csv")
# Load result in

inulin_degraders <- read.csv("data/16S human cohort/taxa_filtered_annotated.csv", row.names = 1, na.strings = "") %>%
  filter(!is.na(vmh_match_species)) %>%
  rownames()
# 13 taxa remain with a species level match, 31 with genus level

clr2 <- clr %>%
  dplyr::select(c(inulin_degraders, group))

# Create empty objects to be loaded with output statistics
importanceDf2 <- data.frame()
accuracy2 <- c()
kappa2 <- c()
pvals2 <- c()
auc2 <- c()
roc.data2 <- data.frame()

# Iterate through the random forest model 100 times
for (i in 1:100){
  # create training/testing split
  data <- clr2
  rows <- sample(nrow(clr2)) # randomly shuffle rows of the dataset to create additional variation between iterations
  data <- data[rows,]
  
  # Train the model
  fitControl <- trainControl(method = "LOOCV",
                             summaryFunction = prSummary,
                             classProbs = T,
                             savePredictions = TRUE) 
  rf <- train(group ~ ., data = data,
              method = "rf",
              trControl = fitControl,
              tuneGrid = expand.grid(.mtry=sqrt(ncol(data)-1)), # default val of sqrt(# features)
              verbose = FALSE)
  
  # Store the scaled importance values
  importances <- varImp(rf)$importance %>% as.matrix %>% t()
  
  # Summarize confusion matrix
  confusion <- confusionMatrix(data = rf$pred$pred, reference = rf$pred$obs,
                               mode = "prec_recall", positive = "Prebiotic")
  
  # Compile resulting metrics
  ## Could pull other values (Mcnemar's Test P-Value, Precision, Recall, F1, etc.) in similar manner
  importanceDf2 <- rbind(importanceDf2, importances)
  accuracy2 <- append(accuracy2, confusion$overall["Accuracy"])
  kappa2 <- append(kappa2, confusion$overall["Kappa"])
  pvals2 <- append(pvals2, confusion$overall["AccuracyPValue"])
  
  # Compile data for plotting ROC curve
  plots <- evalm(rf)
  roc.data.tmp <- data.frame(SENS=plots$roc$data$SENS, FPR=plots$roc$data$FPR,
                             point=1:42, iteration=paste0("iteration", i))
  roc.data2 <- rbind(roc.data2, roc.data.tmp)
  auc2 <- append(auc2, plots$stdres$`Group 1`[13,1])

  # Print out how far through the iterations we are
  print(paste0(i, "/100 iterations complete"))
}

# Generate ROC curve
roc.avg2 <- roc.data2[,1:3] %>%
  gather(key=variable, value=val, 1:2) %>%
  group_by(variable, point) %>%
  summarize(mean=mean(val)) %>%
  spread(key=variable, value=mean)

roc.plot2 <- ggplot(roc.data2, aes(x=FPR, y=SENS, group=iteration)) +
  geom_line(color="red", alpha=0.2) +
  geom_line(data=roc.avg2, aes(x=FPR, y=SENS), inherit.aes = F, color="black", size=1) +
  labs(x="False Positive Rate", y="True Positive Rate") +
  geom_abline(intercept = 0, slope = 1, color="darkgray", linetype = "dashed") +
  annotate(geom="text", x=0.7, y=0.25, label=paste0("Mean AUC = ", round(mean(auc2), 3), "\n",
                                                     "Mean Accuracy = ", round(mean(accuracy2), 3), "\n",
                                                     "Mean Kappa = ", round(mean(kappa2),3), "\n",
                                                     "Mean p(Acc < NIR) = ", round(mean(pvals2), 3)),
           size=2.8)


# calculate average importance across runs and then sort the data
avgImportances2 <- importanceDf2 %>% 
  colMeans() %>% 
  sort(decreasing = TRUE) %>%
  .[1:10] # select top 10 taxa

# link seq numbers to their family taxa name
taxa2 <- tax_table(ps.clr) %>% as.data.frame()
taxa2 <- taxa2[names(avgImportances2),]
taxa2$Species[is.na(taxa2$Species)] <- "sp."
taxa2$binom <- paste0(taxa2$Genus, " ", taxa2$Species, " (", rownames(taxa2), ")")

top10_2 <- data.frame(tax=names(avgImportances2), importance=avgImportances2, binom = taxa2$binom)
top10_2$binom <- factor(top10_2$binom, levels=rev(top10_2$binom))

# Importance plot
importance.plot2 <- ggplot(top10_2, aes(x=importance, y=binom)) +
  geom_point() +
  theme_bw() +
  labs(x="Importance", y=NULL) +
  theme(axis.text.y=element_text(size=6))

# CLR plot
clr.select2 <- clr2[,colnames(clr2) %in% c("group", rownames(top10_2))]
clr.select2 <- cbind(data.frame(participant=rownames(clr.select2)), clr.select2[,c(11,1:10)])
clr.select2 <- clr.select2 %>% 
  gather(key=tax, value=clr, 3:12)
for(r in 1:nrow(clr.select2)) {
  clr.select2$binom[r] <- as.character(top10_2[top10_2$tax == clr.select2$tax[r], "binom"])
}
clr.select2$binom <- factor(clr.select2$binom, levels=rev(top10_2$binom))

clr.summary2 <- clr.select2 %>%
  group_by(group, binom) %>%
  summarize(mean_clr=mean(clr), se=sd(clr)/sqrt(length(clr)))

clr.plot2 <- ggplot(clr.select2, aes(x=clr, y=binom, group=group, color=group)) +
  geom_point(position=position_jitter(height=0.1), alpha=0.5, size=0.5) +
  geom_errorbar(data=clr.summary2, aes(y=binom, xmin=mean_clr, xmax=mean_clr, group=group, color=group),
                size=0.75, width=0.6, inherit.aes = F) +
    geom_errorbar(data=clr.summary2, aes(y=binom, xmin=mean_clr-se, xmax=mean_clr+se, group=group, color=group),
                size=0.75, width=0.3, inherit.aes = F) +
  theme_bw() +
  labs(x="ΔCLR(Relative Abundance)", y=NULL, color="Group") +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())

join.plot2 <- importance.plot2 + clr.plot2 + plot_layout(widths = c(1, 2))

```

# Patch together figure
```{r}
figS3 <- ord_plot / (volcano_1 + volcano_4) / (volcano_mp_baseline + volcano_mp_pre + volcano_mp_pre2) +
  plot_annotation(tag_levels = "A")
#ggsave("plots/figS3.png", figS3, height=8, width=8.1)


```

